### PR TITLE
docstore: test BeforeDo

### PR DESCRIPTION
**Pull request by jba**
Also add BeforeDo support and fix some bugs related to it.

Part of #2260.

### CODE REVIEW

File: internal/docstore/drivertest/drivertest.go [hunk 1]
```diff
@@ -1382,40 +1381,86 @@ func testUnorderedActions(t *testing.T, coll *ds.Collection, revField string) {
 	}
 }
 
-func testAs(t *testing.T, coll *ds.Collection, st AsTest) {
-	docs := []*HighScore{
-		{game3, "steph", 24, date(4, 25), nil},
-		{game3, "mia", 99, date(4, 26), nil},
+// Verify that BeforeDo is invoked, and its as function behaves as expected.
+func testBeforeDo(t *testing.T, newHarness HarnessMaker) {
```
Comment (jba) [inline internal/docstore/drivertest/drivertest.go]: I think this is a more straightforward and comprehensive way to test escape hatches like `BeforeDo` than the way we do it elsewhere, where each driver must implement an `AsTest` interface and write its own tests.

@vangent WDYT?

File: internal/docstore/memdocstore/mem.go [hunk 2]
```diff
@@ -151,6 +154,15 @@ func (c *collection) RunActions(ctx context.Context, actions []*driver.Action, o
 		t.Wait()
 	}
 
+	if opts.BeforeDo != nil {
```
Comment (shantuo) [inline internal/docstore/memdocstore/mem.go]: This doesn't do anything, can we remove it?

File: internal/docstore/drivertest/drivertest.go (hunk 1 above)
Comment (shantuo) [reply internal/docstore/drivertest/drivertest.go]: I think this way you save the work of implementing `AsTest` by writing the `XXXTypes` function.

File: internal/docstore/memdocstore/mem.go (hunk 2 above)
Comment (jba) [reply internal/docstore/memdocstore/mem.go]: It lets the user implement a `BeforeDo` that aborts the execution of the list. But yeah, that is probably not useful. But I need it for this test, so I guess we should leave it? If I take it out, how do I get memdocstore to pass the test?

**Comment by jba** [comment]
PTAL

File: internal/docstore/docstore.go [hunk 3]
```diff
@@ -260,10 +260,14 @@ func (e ActionListError) Unwrap() error {
 	return nil
 }
 
-// BeforeDo takes a callback function that will be called before the ActionList
-// is executed by the underlying provider's action functionality. The callback
-// takes a parameter, asFunc, that converts its argument to provider-specific
-// types. See https://gocloud.dev/concepts/as/ for background information.
+// BeforeDo takes a callback function that will be called before the ActionList is
+// executed by the underlying provider. It may be invoked multiple times for a
+// single call to ActionList.Do. If the callback returns an error, ActionList.Do
```
Comment (vangent) [inline internal/docstore/docstore.go]: When would it be invoked multiple times? Consider explaining this.

File: internal/docstore/drivertest/drivertest.go [hunk 4]
```diff
@@ -175,6 +172,8 @@ func RunConformanceTests(t *testing.T, newHarness HarnessMaker, ct CodecTester,
 	t.Run("DeleteQuery", func(t *testing.T) { withTwoKeyCollection(t, newHarness, testDeleteQuery) })
 	t.Run("UpdateQuery", func(t *testing.T) { withTwoKeyCollection(t, newHarness, testUpdateQuery) })
 
+	t.Run("BeforeDo", func(t *testing.T) { testBeforeDo(t, newHarness) })
```
Comment (vangent) [inline internal/docstore/drivertest/drivertest.go]: I suspect you will end up wanting to combine this with other `As` tests, but you can worry about that later.

File: internal/docstore/drivertest/drivertest.go (hunk 1 above)
Comment (vangent) [reply internal/docstore/drivertest/drivertest.go]: It seems fine. You lose the ability to do an end-to-end test of `As` -- for example, `gcsblob` uses `As` to set a field that's not supported by `blob` in the `BeforeWrite`, and then reads it back using the `As` for read. If you don't want to do that, this is more succinct.

File: internal/docstore/drivertest/drivertest.go [hunk 5]
```diff
@@ -53,6 +54,12 @@ type Harness interface {
 	// that the revision field should be drivertest.AlternateRevisionField.
 	MakeAlternateRevisionFieldCollection(context.Context) (driver.Collection, error)
 
+	// BeforeDoTypes should return a list of values whose types are valid for the as
+	// function given to BeforeDo. For example, if the provider converts Get actions
```
Comment (vangent) [inline internal/docstore/drivertest/drivertest.go]: It should be clearer from this whether `As` should work for all of the types, or just one of them. The test should probably support both, as there are cases where a single `As` will support multiple types (not one or the other, but both).

File: internal/docstore/dynamodocstore/dynamo.go [hunk 6]
```diff
@@ -30,7 +30,8 @@
 //
 // dynamodocstore exposes the following types for As:
 //  - Collection.As: *dynamodb.DynamoDB
-//  - ActionList.BeforeDo: *dynamodb.BatchGetItemsInput or *dynamodb.TransactWriteItemsInput
+//  - ActionList.BeforeDo: *dynamodb.BatchGetItemInput, *dynamodb.PutItemInput, *dynamodb.DeleteItemInput,
```
Comment (vangent) [inline internal/docstore/dynamodocstore/dynamo.go]: I think this should say "or" to make it more clear that each call will support one of these, not all.

File: internal/docstore/drivertest/drivertest.go (hunk 5 above)
Comment (jba) [reply internal/docstore/drivertest/drivertest.go]: I try BeforeDo on all six action types, and each action type must support at least one of the BeforeDo types.

I thought about making this more fine-grained, but it's tricky. A Get may have one type, a Put another, and an Update two others. It's going to get messy to specify all that. I figured this was a good start.

How about I put a TODO to make it more action-specific?

File: internal/docstore/drivertest/drivertest.go (hunk 4 above)
Comment (jba) [reply internal/docstore/drivertest/drivertest.go]: Agreed.

File: internal/docstore/docstore.go (hunk 3 above)
Comment (jba) [reply internal/docstore/docstore.go]: Done.

**Comment by jba** [comment]
PTAL.

File: internal/docstore/dynamodocstore/dynamo.go (hunk 6 above)
Comment (vangent) [reply internal/docstore/dynamodocstore/dynamo.go]: please fix

Comment (jba) [reply internal/docstore/dynamodocstore/dynamo.go]: Done.

File: internal/docstore/drivertest/drivertest.go (hunk 5 above)
Comment (jba) [reply internal/docstore/drivertest/drivertest.go]: I added the TODO.

**Pull request merged**